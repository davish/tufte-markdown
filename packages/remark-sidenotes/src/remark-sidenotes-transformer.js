import getSlug from 'speakingurl'
import visit from 'unist-util-visit'
import select from 'unist-util-select'
import toHAST from 'mdast-util-to-hast'
import toHTML from 'hast-util-to-html'

const MARGINNOTE_SYMBOL = '{-}'

function sidenotes() {
  return transformer
}

const generateLabel = (isMarginNote, title) =>
  `${isMarginNote ? 'md' : 'sd'}-${getSlug(title, { truncate: 20 })}`

const getReplacement = ({ isMarginNote, noteHTML }) => {
  const label = generateLabel(isMarginNote, noteHTML)
  const labelCls = `margin-toggle ${isMarginNote ? '' : 'sidenote-number'}`
  const labelSymbol = isMarginNote ? '&#8853;' : ''
  const noteTypeCls = isMarginNote ? 'marginnote' : 'sidenote'

  return [
    {
      type: 'html',
      value: `<label for="${label}" class="${labelCls}">${labelSymbol}</label>`,
    },
    {
      type: 'html',
      value: `<input type="checkbox" id="${label}" class="margin-toggle" />`,
    },
    {
      type: 'html',
      value: `<span class="${noteTypeCls}">`,
    },
    {
      type: 'html',
      value: noteHTML,
    },
    {
      type: 'html',
      value: '</span>',
    },
  ]
}

const coerceToHtml = nodeArray =>
  nodeArray.map(node => toHTML(toHAST(node))).join('') || ''

const extractNoteFromHtml = note => {
  const matches = note.match(/(\s+)*({-})*\s*((.|\n)+)/)

  return {
    isMarginNote: matches[2] === MARGINNOTE_SYMBOL,
    noteHTML: matches[3],
  }
}

function transformer(tree) {
  // "Regular" Sidenotes/Marginnotes consisting of a reference and a definition
  // Syntax for Sidenotes [^<number>] and somewhere else [^<number]: <markdown>
  // Syntax for Marginnotes [^<descriptor>] and somewhere else [^<descriptor]: {-}
  visit(tree, 'footnoteReference', (node, index, parent) => {
    const target = select(
      tree,
      `footnoteDefinition[identifier=${node.identifier}]`
    )

    if (!target) throw new Error('No coresponding note found')

    const notesAst =
      target[0].children.length && target[0].children[0].type === 'paragraph'
        ? target[0].children[0].children
        : target[0].children

    const nodeDetail = extractNoteFromHtml(coerceToHtml(notesAst))

    parent.children.splice(index, 1, ...getReplacement(nodeDetail))
  })

  visit(tree, 'footnoteDefinition', (node, index, parent) => {
    parent.children.splice(index, 1)
  })

  // "Inline" Sidenotes which do not have two parts
  // Syntax: [^{-} <markdown>]
  visit(tree, 'footnote', (node, index, parent) => {
    console.log()
    const notesAst = node.children
    const nodeDetail = extractNoteFromHtml(coerceToHtml(notesAst))

    parent.children.splice(index, 1, ...getReplacement(nodeDetail))
  })
}

export default sidenotes
