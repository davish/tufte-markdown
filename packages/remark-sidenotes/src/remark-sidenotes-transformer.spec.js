import u from 'unist-builder'
import select from 'unist-util-select'
import HTML from 'html-parse-stringify'
import { transformer } from './remark-sidenotes-transformer'

const clone = obj => Object.assign({}, obj)

describe('Sidenotes', () => {
  describe('with target not as its own paragraph', () => {
    let ast
    let subject
    let result
    beforeEach(() => {
      ast = u('root', [
        u('paragraph', [
          u('text', 'Lorem ipsum dolor sed amet'),
          u('footnoteReference', {
            identifier: '^1',
          }),
          u('text', 'dolores adam turing and the gang'),
          u('emphasis', 'rocked'),
          u('text', 'their life'),
        ]),
        u('paragraph', [
          u('footnoteDefinition', { identifier: '^1' }, [
            u('text', 'This is the footnote for identifier ^1'),
            u('emphasis', [u('text', 'This is strong')]),
            u(
              'link',
              { url: 'http://example.org', title: 'Just an example' },
              'Link to example.org'
            ),
          ]),
        ]),
      ])

      subject = clone(ast)
      result = transformer(subject)
    })

    it('should match snapshot', () => {
      expect(result).toMatchSnapshot()
    })

    describe('label-element', () => {
      it('should replace the `footnoteReference` with a `label`', () => {
        expect(select(result, 'html[value^=<label]')).toHaveLength(1)
        expect(select(result, 'footnoteReference')).toHaveLength(0)
      })

      it('should generate the correct identifier for the label', () => {
        const labelNode = HTML.parse(
          select(result, 'html[value^=<label]')[0].value
        )[0]

        expect(labelNode.attrs.for).toEqual('sd-this-is-the-footnote')
      })

      it('should generate the correct css-class for the label', () => {
        const labelNode = HTML.parse(
          select(result, 'html[value^=<label]')[0].value
        )[0]
        expect(labelNode.attrs.class).toEqual('margin-toggle sidenote-number')
      })

      it('should not generate any child notes for the label', () => {
        const labelNode = HTML.parse(
          select(result, 'html[value^=<label]')[0].value
        )[0]
        expect(labelNode.children).toHaveLength(0)
      })
    })

    describe('input-element', () => {
      it('should replace the `footnoteReference` with an `input`', () => {
        expect(select(result, 'html[value^=<input]')).toHaveLength(1)
        expect(select(result, 'footnoteReference')).toHaveLength(0)
      })

      it('should be of type="checkbox"', () => {
        const inputNode = HTML.parse(
          select(result, 'html[value^=<input]')[0].value
        )[0]

        expect(inputNode.attrs.type).toEqual('checkbox')
      })

      it('should generate the correct id', () => {
        const inputNode = HTML.parse(
          select(result, 'html[value^=<input]')[0].value
        )[0]

        expect(inputNode.attrs.id).toEqual('sd-this-is-the-footnote')
      })

      it('should have the correct css-class', () => {
        const inputNode = HTML.parse(
          select(result, 'html[value^=<input]')[0].value
        )[0]

        expect(inputNode.attrs.class).toEqual('margin-toggle')
      })
    })

    describe('sidenote content', () => {
      it('should replace the `footnoteDefinition` with a span element', () => {
        expect(
          select(result, 'html[value^=<span class="sidenote]')
        ).toHaveLength(1)
        expect(select(result, 'footnoteDefinition')).toHaveLength(0)
      })
    })
  })

  describe('with target as paragraph', () => {
    let ast
    let subject
    let result
    beforeEach(() => {
      ast = u('root', [
        u('paragraph', [
          u('text', 'Lorem ipsum dolor sed amet'),
          u('footnoteReference', {
            identifier: '^1',
          }),
          u('text', 'dolores adam turing and the gang'),
          u('emphasis', 'rocked'),
          u('text', 'their life'),
        ]),
        u('paragraph', [
          u('footnoteDefinition', { identifier: '^1' }, [
            u('paragraph', [
              u('text', 'This is the footnote for identifier ^1'),
              u('emphasis', [u('text', 'This is strong')]),
              u(
                'link',
                { url: 'http://example.org', title: 'Just an example' },
                'Link to example.org'
              ),
            ]),
          ]),
        ]),
      ])

      subject = clone(ast)
      result = transformer(subject)
    })

    it('should match the snapshot', () => {
      expect(result).toMatchSnapshot()
    })

    it('should remove the wrapping paragraph from the note-content', () => {
      const note = select(
        result,
        'html[value^=<span class="sidenote] + html'
      )[0]
      expect(note.value).not.toMatch(/^<p>/)
    })
  })

  describe('with no matching `noteDefinition`', () => {
    let ast
    let subject
    beforeEach(() => {
      ast = u('root', [
        u('paragraph', [
          u('text', 'Lorem ipsum dolor sed amet'),
          u('footnoteReference', {
            identifier: '^1',
          }),
          u('text', 'dolores adam turing and the gang'),
          u('emphasis', 'rocked'),
          u('text', 'their life'),
        ]),
      ])

      subject = clone(ast)
    })

    it('should throw an error', () => {
      function transformAst() {
        transformer(subject)
      }
      expect(transformAst).toThrowErrorMatchingSnapshot()
    })
  })
})

describe('Marginnotes', () => {
  describe('with margin-note-indicator inline in text', () => {
    let ast
    let subject
    let result
    beforeEach(() => {
      ast = u('root', [
        u('paragraph', [
          u('text', 'Lorem ipsum dolor sed amet'),
          u('footnoteReference', {
            identifier: '^blue',
          }),
          u('text', 'dolores adam turing and the gang'),
          u('emphasis', 'rocked'),
          u('text', 'their life'),
        ]),
        u('paragraph', [
          u('footnoteDefinition', { identifier: '^blue' }, [
            u('text', '{-} This is the footnote for identifier ^blue'),
            u('emphasis', [u('text', 'This is strong')]),
            u(
              'link',
              { url: 'http://example.org', title: 'Just an example' },
              'Link to example.org'
            ),
          ]),
        ]),
      ])

      subject = clone(ast)
      result = transformer(subject)
    })

    it('should match snapshot', () => {
      expect(result).toMatchSnapshot()
    })

    describe('label-element', () => {
      it('should replace the `footnoteReference` with a `label`', () => {
        expect(select(result, 'html[value^=<label]')).toHaveLength(1)
        expect(select(result, 'footnoteReference')).toHaveLength(0)
      })

      it('should generate the correct identifier for the label', () => {
        const labelNode = HTML.parse(
          select(result, 'html[value^=<label]')[0].value
        )[0]

        expect(labelNode.attrs.for).toEqual('md-this-is-the-footnote')
      })

      it('should generate the correct css-class for the label', () => {
        const labelNode = HTML.parse(
          select(result, 'html[value^=<label]')[0].value
        )[0]
        expect(labelNode.attrs.class).toContain('margin-toggle')
      })

      it('should not generate any child notes for the label', () => {
        const labelNode = HTML.parse(
          select(result, 'html[value^=<label]')[0].value
        )[0]
        expect(labelNode.children[0].content).toContain('&#8853;')
        expect(labelNode.children).toHaveLength(1)
      })
    })

    describe('input-element', () => {
      it('should replace the `footnoteReference` with an `input`', () => {
        expect(select(result, 'html[value^=<input]')).toHaveLength(1)
        expect(select(result, 'footnoteReference')).toHaveLength(0)
      })

      it('should be of type="checkbox"', () => {
        const inputNode = HTML.parse(
          select(result, 'html[value^=<input]')[0].value
        )[0]

        expect(inputNode.attrs.type).toEqual('checkbox')
      })

      it('should generate the correct id', () => {
        const inputNode = HTML.parse(
          select(result, 'html[value^=<input]')[0].value
        )[0]

        expect(inputNode.attrs.id).toEqual('md-this-is-the-footnote')
      })

      it('should have the correct css-class', () => {
        const inputNode = HTML.parse(
          select(result, 'html[value^=<input]')[0].value
        )[0]

        expect(inputNode.attrs.class).toEqual('margin-toggle')
      })
    })

    describe('sidenote content', () => {
      it('should replace the `footnoteDefinition` with a span element', () => {
        expect(
          select(result, 'html[value^=<span class="marginnote]')
        ).toHaveLength(1)
        expect(select(result, 'footnoteDefinition')).toHaveLength(0)
      })
    })
  })

  describe('with margin-note-indicator as own node', () => {
    let ast
    let subject
    let result
    beforeEach(() => {
      ast = u('root', [
        u('paragraph', [
          u('text', 'Lorem ipsum dolor sed amet'),
          u('footnoteReference', {
            identifier: '^red',
          }),
          u('text', 'dolores adam turing and the gang'),
          u('emphasis', 'rocked'),
          u('text', 'their life'),
        ]),
        u('paragraph', [
          u('footnoteDefinition', { identifier: '^red' }, [
            u('text', '{-} \n'),
            u('text', 'This is the footnote for identifier ^red'),
            u('emphasis', [u('text', 'This is strong')]),
            u(
              'link',
              { url: 'http://example.org', title: 'Just an example' },
              'Link to example.org'
            ),
          ]),
        ]),
      ])

      subject = clone(ast)
      result = transformer(subject)
    })

    it('should match snapshot', () => {
      expect(result).toMatchSnapshot()
    })
  })

  describe('with no matching `noteDefinition`', () => {
    let ast
    let subject
    beforeEach(() => {
      ast = u('root', [
        u('paragraph', [
          u('text', 'Lorem ipsum dolor sed amet'),
          u('footnoteReference', {
            identifier: '^blue',
          }),
          u('text', 'dolores adam turing and the gang'),
          u('emphasis', 'rocked'),
          u('text', 'their life'),
        ]),
      ])

      subject = clone(ast)
    })

    it('should throw an error', () => {
      function transformAst() {
        transformer(subject)
      }
      expect(transformAst).toThrowErrorMatchingSnapshot()
    })
  })
})

describe('Inline-Marginnotes', () => {
  let ast
  let subject
  let result
  beforeEach(() => {
    ast = u('root', [
      u('paragraph', [
        u('tufteFigure', [
          u('footnote', [
            u('text', '{-} This is the footnote for identifier ^blue'),
            u('emphasis', [u('text', 'This is strong')]),
            u(
              'link',
              { url: 'http://example.org', title: 'Just an example' },
              'Link to example.org'
            ),
          ]),
          u('image', {
            title: 'bravo',
            url: 'http://example.com',
            alt: 'alpha',
          }),
        ]),
      ]),
    ])

    subject = clone(ast)
    result = transformer(subject)
  })

  it('should match snapshot', () => {
    expect(result).toMatchSnapshot()
  })
})
