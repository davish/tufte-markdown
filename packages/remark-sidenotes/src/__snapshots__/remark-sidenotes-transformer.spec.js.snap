// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline-Marginnotes should match snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "html",
              "value": "<label for=\\"md-this-is-the-footnote\\" class=\\"margin-toggle \\">&#8853;</label>",
            },
            Object {
              "type": "html",
              "value": "<input type=\\"checkbox\\" id=\\"md-this-is-the-footnote\\" class=\\"margin-toggle\\" />",
            },
            Object {
              "type": "html",
              "value": "<span class=\\"marginnote\\">",
            },
            Object {
              "type": "html",
              "value": "This is the footnote for identifier ^blue<em>This is strong</em><a href=\\"http://example.org\\" title=\\"Just an example\\"></a>",
            },
            Object {
              "type": "html",
              "value": "</span>",
            },
            Object {
              "alt": "alpha",
              "title": "bravo",
              "type": "image",
              "url": "http://example.com",
            },
          ],
          "type": "tufteFigure",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Marginnotes with margin-note-indicator as own node should match snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Lorem ipsum dolor sed amet",
        },
        Object {
          "type": "html",
          "value": "<label for=\\"md-this-is-the-footnote\\" class=\\"margin-toggle \\">&#8853;</label>",
        },
        Object {
          "type": "html",
          "value": "<input type=\\"checkbox\\" id=\\"md-this-is-the-footnote\\" class=\\"margin-toggle\\" />",
        },
        Object {
          "type": "html",
          "value": "<span class=\\"marginnote\\">",
        },
        Object {
          "type": "html",
          "value": "This is the footnote for identifier ^red<em>This is strong</em><a href=\\"http://example.org\\" title=\\"Just an example\\"></a>",
        },
        Object {
          "type": "html",
          "value": "</span>",
        },
        Object {
          "type": "text",
          "value": "dolores adam turing and the gang",
        },
        Object {
          "type": "emphasis",
          "value": "rocked",
        },
        Object {
          "type": "text",
          "value": "their life",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Marginnotes with margin-note-indicator inline in text should match snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Lorem ipsum dolor sed amet",
        },
        Object {
          "type": "html",
          "value": "<label for=\\"md-this-is-the-footnote\\" class=\\"margin-toggle \\">&#8853;</label>",
        },
        Object {
          "type": "html",
          "value": "<input type=\\"checkbox\\" id=\\"md-this-is-the-footnote\\" class=\\"margin-toggle\\" />",
        },
        Object {
          "type": "html",
          "value": "<span class=\\"marginnote\\">",
        },
        Object {
          "type": "html",
          "value": "This is the footnote for identifier ^blue<em>This is strong</em><a href=\\"http://example.org\\" title=\\"Just an example\\"></a>",
        },
        Object {
          "type": "html",
          "value": "</span>",
        },
        Object {
          "type": "text",
          "value": "dolores adam turing and the gang",
        },
        Object {
          "type": "emphasis",
          "value": "rocked",
        },
        Object {
          "type": "text",
          "value": "their life",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Marginnotes with no matching \`noteDefinition\` should throw an error 1`] = `"No coresponding note found"`;

exports[`Sidenotes with no matching \`noteDefinition\` should throw an error 1`] = `"No coresponding note found"`;

exports[`Sidenotes with target as paragraph should match the snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Lorem ipsum dolor sed amet",
        },
        Object {
          "type": "html",
          "value": "<label for=\\"sd-this-is-the-footnote\\" class=\\"margin-toggle sidenote-number\\"></label>",
        },
        Object {
          "type": "html",
          "value": "<input type=\\"checkbox\\" id=\\"sd-this-is-the-footnote\\" class=\\"margin-toggle\\" />",
        },
        Object {
          "type": "html",
          "value": "<span class=\\"sidenote\\">",
        },
        Object {
          "type": "html",
          "value": "This is the footnote for identifier ^1<em>This is strong</em><a href=\\"http://example.org\\" title=\\"Just an example\\"></a>",
        },
        Object {
          "type": "html",
          "value": "</span>",
        },
        Object {
          "type": "text",
          "value": "dolores adam turing and the gang",
        },
        Object {
          "type": "emphasis",
          "value": "rocked",
        },
        Object {
          "type": "text",
          "value": "their life",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Sidenotes with target not as its own paragraph should match snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Lorem ipsum dolor sed amet",
        },
        Object {
          "type": "html",
          "value": "<label for=\\"sd-this-is-the-footnote\\" class=\\"margin-toggle sidenote-number\\"></label>",
        },
        Object {
          "type": "html",
          "value": "<input type=\\"checkbox\\" id=\\"sd-this-is-the-footnote\\" class=\\"margin-toggle\\" />",
        },
        Object {
          "type": "html",
          "value": "<span class=\\"sidenote\\">",
        },
        Object {
          "type": "html",
          "value": "This is the footnote for identifier ^1<em>This is strong</em><a href=\\"http://example.org\\" title=\\"Just an example\\"></a>",
        },
        Object {
          "type": "html",
          "value": "</span>",
        },
        Object {
          "type": "text",
          "value": "dolores adam turing and the gang",
        },
        Object {
          "type": "emphasis",
          "value": "rocked",
        },
        Object {
          "type": "text",
          "value": "their life",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;
